src/npctalk.cpp

```diff
@@ -1970,27 +1970,25 @@ void talk_effect_fun_t::set_u_sell_item( const std::string &item_name, int cost,
     function = [item_name, cost, count]( const dialogue & d ) {
         npc &p = *d.beta;
         player &u = *d.alpha;
-        item old_item = item( item_name, calendar::turn );
-        if( old_item.count_by_charges() && u.has_charges( item_name, count ) ) {
-            u.use_charges( item_name, count );
-            old_item.mod_charges( count - 1 );
-            p.i_add( old_item );
+        if( item::count_by_charges( item_name ) && u.has_charges( item_name, count ) ) {
+            for( const item &it : u.use_charges( item_name, count ) ) {
+                p.i_add( it );
+            }
         } else if( u.has_amount( item_name, count ) ) {
-            u.use_amount( item_name, count );
-            for( int i_cnt = 0; i_cnt < count; i_cnt++ ) {
-                p.i_add( old_item );
+            for( const item &it : u.use_amount( item_name, count ) ) {
+                p.i_add( it );
             }
         } else {
             //~ %1$s is a translated item name
-            popup( _( "You don't have a %1$s!" ), old_item.tname() );
+            popup( _( "You don't have a %1$s!" ), item_name );
             return;
         }
         if( count == 1 ) {
             //~ %1%s is the NPC name, %2$s is an item
-            popup( _( "You give %1$s a %2$s." ), p.name, old_item.tname() );
+            popup( _( "You give %1$s a %2$s." ), p.name, item_name );
         } else {
             //~ %1%s is the NPC name, %2$d is a number of items, %3$s are items
-            popup( _( "You give %1$s %2$d %3$s." ), p.name, count, old_item.tname() );
+            popup( _( "You give %1$s %2$d %3$s." ), p.name, count, item_name );
```

Comment:
```suggestion
            popup( _( "You give %1$s %2$d %3$s." ), item::nname( item_name, count ) );
```
The count parameter gets the proper plural form. And the variable `item_name` is in fact the item id. It's not something we want to display at all.

Reply:
Alright, I've made another commit that I think uses item::nname correctly - the popup message looks correct, but the NPC still doesn't get an inhaler until I save and reload the game.